osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyUserRankCounts.cs

```diff
@@ -0,0 +1,161 @@
+                    { ScoreRank.S, 0 },
+                    { ScoreRank.SH, 0 },
+                    { ScoreRank.X, 0 },
+                    { ScoreRank.XH, 0 },
+                };
+
+                if (Verbose)
+                    Console.WriteLine($"Processing user {userId} ({scores.Count()} scores)..");
+
+                IEnumerable<SoloScore?> maxScoresByBeatmap = scores.GroupBy(s => s.beatmap_id).Select(g => g.MaxBy(s => s.total_score));
+
+                foreach (var score in maxScoresByBeatmap)
+                {
+                    if (score == null)
+                        continue;
+
+                    switch (score.rank)
+                    {
+                        case ScoreRank.A:
+                            counts[ScoreRank.A]++;
+                            break;
+
+                        case ScoreRank.S:
+                            counts[ScoreRank.S]++;
+                            break;
+
+                        case ScoreRank.SH:
+                            counts[ScoreRank.SH]++;
+                            break;
+
+                        case ScoreRank.X:
+                            counts[ScoreRank.X]++;
+                            break;
+
+                        case ScoreRank.XH:
+                            counts[ScoreRank.XH]++;
+                            break;
+                    }
+                }
+
+                var userStats = await DatabaseHelper.GetUserStatsAsync(userId, RulesetId, db, transaction);
+
+                if (userStats == null)
+                    return;
+
+                bool userHasCorrectCounts =
+                    userStats.a_rank_count == counts[ScoreRank.A] &&
+                    userStats.s_rank_count == counts[ScoreRank.S] &&
+                    userStats.sh_rank_count == counts[ScoreRank.SH] &&
+                    userStats.x_rank_count == counts[ScoreRank.X] &&
+                    userStats.xh_rank_count == counts[ScoreRank.XH];
+
+                if (!userHasCorrectCounts)
+                {
+                    if (Verbose)
+                    {
+                        Console.WriteLine($"Fixing incorrect counts for {userId}");
+                        Console.WriteLine($"a:  {userStats.a_rank_count} ->  {counts[ScoreRank.A]}");
+                        Console.WriteLine($"s:  {userStats.s_rank_count} ->  {counts[ScoreRank.S]}");
+                        Console.WriteLine($"sh: {userStats.sh_rank_count} ->  {counts[ScoreRank.SH]}");
+                        Console.WriteLine($"x:  {userStats.x_rank_count} ->  {counts[ScoreRank.X]}");
+                        Console.WriteLine($"xh: {userStats.xh_rank_count} ->  {counts[ScoreRank.XH]}");
+                        Console.WriteLine();
+                    }
+
+                    if (!DryRun)
+                        await DatabaseHelper.UpdateUserStatsAsync(userStats, db, transaction);
+                }
+
+                await transaction.CommitAsync(cancellationToken);
```

Comment:
Bit unfortunate that even in dry run mode this is gonna hold a transaction open for the entire duration of the operation even despite issuing no sql commands. Also a commit in a dry run mode is a tad scary.

In principle all of this is safe and good as written, but I do wonder if things can't be reshuffled so that dry run never takes out a txn in the first place.

Reply:
I can just make it rollback if in dryrun if that is enough to make you feel more at ease?

Reply:
If it works for you it works for me. That part wasn't the one I was worried about so much, the part that worries me more is unnecessary transaction contention when running this in dry run (although the devil's advocate take on that is that *incurring* this transaction contention will be useful to check if things don't fall over too hard on the *real* run).

Reply:
Yeah, I'd want the transaction to remain even in a dry run for that purpose too.